src/storage/src/source/mod.rs

```diff
@@ -806,169 +866,640 @@ where
             source_metrics.operator_scheduled_counter.inc();
 
             let mut context = Context::from_waker(&waker);
 
             let timer = Instant::now();
 
-            while let Poll::Ready(Some(event)) = source_reader.as_mut().poll_next(&mut context) {
-                match event {
-                    Event::Progress(upper) => {
-                        let ts = upper.as_option().cloned().unwrap_or(Timestamp::MAX);
-                        for partition_metrics in source_metrics.partition_metrics.values_mut() {
-                            partition_metrics.closed_ts.set(ts);
-                        }
+            // We just use an advancing number for our capability. No one cares
+            // about what this actually is, downstream.
+            let mut batch_counter = 0;
+
+            while let Poll::Ready(Some(update)) = source_reader.as_mut().poll_next(&mut context) {
+                let (messages, non_definite_errors, unconsumed_partitions, source_upper) =
+                    if let Some(update) = update {
+                        update
+                    } else {
+                        trace!("source_reader({id}) {worker_id}/{worker_count}: is terminated");
+                        // We will never produce more data, clear our capabilities to
+                        // communicate this downstream.
+                        cap_set.downgrade(&[]);
+                        return;
+                    };
+
+                trace!(
+                    "create_source_raw({id}) {worker_id}/{worker_count}: message_batch.len(): {:?}",
+                    messages.len()
+                );
+                trace!(
+                    "create_source_raw({id}) {worker_id}/{worker_count}: source_upper: {:?}",
+                    source_upper
+                );
+
+                // We forward only the partitions that we are responsible for to
+                // the remap operator.
+                let source_upper_summary = SourceUpperSummary {
+                    source_upper: source_upper.clone(),
+                };
+
+                // Pull the upper to `max` for partitions that we are not
+                // responsible for. That way, the downstream reclock operator
+                // can correcly decide when a reclocked timestamp is closed.
+                // We basically take those partitions "out of the calculation".
+                let mut extended_source_upper = source_upper.clone();
+                extended_source_upper.extend(
+                    unconsumed_partitions
+                        .into_iter()
+                        .map(|pid| (pid, MzOffset { offset: u64::MAX })),
+                );
+
+                let non_definite_errors = non_definite_errors
+                    .into_iter()
+                    .map(|e| SourceError {
+                        source_id: id,
+                        error: e.inner,
+                    })
+                    .collect_vec();
```

Comment:
I wonder how much additional allocation we are doing here; This may represent additional slowdown for the pg source, but should be a relatively fixed cost per batch, i think